### ISSUE TITLE
HttpLoggingInterceptor logging at level.BODY does not allow streaming requests

### ISSUE THREAD

**Issue by foxish** [square/okhttp#3157]
What kind of issue is this?

 - [ ] Question. This issue tracker is not the place for questions. If you want to ask how to do
       something, or to understand why something isn't working the way you expect it to, use Stack
       Overflow. https://stackoverflow.com/questions/tagged/okhttp

 - [x] Bug report. If you’ve found a bug, spend the time to write a failing test. Bugs with tests
       get fixed. Here’s an example: https://gist.github.com/swankjesse/981fcae102f513eb13ed

 - [ ] Feature Request. Start by telling us what problem you’re trying to solve. Often a solution
       already exists! Don’t send pull requests to implement new features without first getting our
       support. Sometimes we leave features out on purpose to keep the project small.

```
package blah;

import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.concurrent.TimeUnit;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.logging.HttpLoggingInterceptor;
import okhttp3.logging.HttpLoggingInterceptor.Level;
import okio.BufferedSource;
import org.apache.commons.lang3.builder.ReflectionToStringBuilder;
import org.apache.commons.lang3.builder.ToStringStyle;

public class t {
  public static void main(String[] args) throws IOException {
    HttpLoggingInterceptor logger = new HttpLoggingInterceptor();
    logger.setLevel(Level.BODY);

    OkHttpClient client = new OkHttpClient().newBuilder().readTimeout(0, TimeUnit.MILLISECONDS)
        .addNetworkInterceptor(logger)
        .build();
    URL url = null;
    try {
      url = new URL("http://server:port/path");
    } catch (MalformedURLException e) {
      e.printStackTrace();
    }
    HttpUrl.Builder httpUrlBuilder = HttpUrl.get(url).newBuilder();
    httpUrlBuilder
        .addQueryParameter("resourceVersion", "3266924")
        .addQueryParameter("watch", "true");

    Request request = new Request.Builder()
        .get()
        .url(httpUrlBuilder.build())
        .build();

    client.newCall(request).enqueue(new Callback() {
      public void onFailure(Call call, IOException e) {
        System.out.println("Fail");
      }

      public void onResponse(Call call, Response response) throws IOException {
        BufferedSource source = response.body().source();
        while (!source.exhausted()) {
          String line = source.readUtf8LineStrict();
          System.out.println(line);
        }
      }
    });
  }
}

```

I cannot stream responses if I set `logger.setLevel(Level.BODY);`, and it stops streaming any responses and gets stuck with:

```
Feb 07, 2017 2:38:53 PM okhttp3.internal.platform.Platform log
INFO: Content-Type: application/json
Feb 07, 2017 2:38:53 PM okhttp3.internal.platform.Platform log
INFO: Date: Tue, 07 Feb 2017 22:38:53 GMT
Feb 07, 2017 2:38:53 PM okhttp3.internal.platform.Platform log
INFO: Transfer-Encoding: chunked
```

However, if I set it to `Level.BASIC`, it works as expected. I think behaviorally, I wouldn't expect a logging interceptor to change how the actual response behaves.

**Comment by swankjesse** [comment]
Our logging interceptor is intended to help in debugging. If you want to stream and log you’ll be better off building your own logging interceptor, or by using a sophisticated tool like Chuck.

**Comment by Mkohm** [comment]
A possible workaround so that you still can log with BODY level for other network requests is to clear the network interceptor list for the streaming requests. Just do this for every streaming requests.

```
okHttpClient.newBuilder()
            .apply {
                networkInterceptors().clear()
            }
            .build()
```

**Comment by TomerAberbach** [comment]
@yschimke curious if OkHttp has a different opinion on this issue nowadays, with streaming being much more popular due to LLMs?

We [encountered this in the openai-java SDK](https://github.com/openai/openai-java/issues/455), and I'm happy to implement my own interceptor if that's the recommendation, but I wonder if streaming responses should be detected (based on content type, e.g. text-event/stream, JSONL, etc.) and elided similarly to non-UTF8 response bodies?

e.g. right now y'all do this for non-UTF8:
```
<-- END HTTP (binary ${buffer.size}-byte body omitted)
```

I'm imagining the following for streaming:
```
<-- END HTTP (streaming response body omitted)
```

This is what I will likely do if I have to implement it myself. I'm also happy to send a PR if this sounds like a good idea to you!

**Comment by yschimke** [comment]
That sounds like a useful feature to model. But presumably showing each chunk?

Cc @swankjesse thoughts?

**Comment by TomerAberbach** [comment]
> But presumably showing each chunk?

If you think that's feasible / workable, then I would not object.

If that's hard, then I'm fine without that. Most important thing is to not buffer streaming.

**Comment by TomerAberbach** [comment]
Thanks for reopening! Let me know if you'd like me to make a PR or if you're planning to fix. Either way works for me

**Comment by yschimke** [comment]
@TomerAberbach Yep, PR welcome - can you add a new constant such as Level.STREAMING_BODY.

**Comment by yschimke** [comment]
From the open ai issue

> When I enabled OPENAI_LOG=debug, all stream operations failed and could only be synchronized and returned at once.

**Comment by yschimke** [comment]
@TomerAberbach not sure if you made a start. But I had some time on a rainy weekend, so put up a PR.

https://github.com/square/okhttp/pull/8993

It may not land, but for discussion. I think it's complex enough that it shows it's hard for others to implement with library support (OkHttp or some extra lib).
